skipchain/skipchain.go

```diff
@@ -1360,22 +1349,76 @@ func (s *Service) startBFT(proto string, origRoster, newRoster *onet.Roster, msg
 	}
 }
 
-// PropagateSkipBlock will save a new SkipBlock
-func (s *Service) propagateSkipBlock(msg network.Message) {
-	sbs, ok := msg.(*PropagateSkipBlocks)
+// PropagateGenesisHandler will save a new SkipBlock
+func (s *Service) propagateGenesisHandler(msg network.Message) {
+	pg, ok := msg.(*PropagateGenesis)
 	if !ok {
 		log.Error("Couldn't convert to slice of SkipBlocks")
 		return
 	}
-	for _, sb := range sbs.SkipBlocks {
-		if !s.blockIsFriendly(sb) {
-			log.Lvlf2("%s: block is not friendly: %x", s.ServerIdentity(), sb.Hash)
+
+	s.db.Store(pg.Genesis)
+
+	if !pg.Genesis.ParentBlockID.IsNull() {
+		sb := s.db.GetByID(pg.Genesis.ParentBlockID)
+		if sb == nil {
+			log.Lvl3("Unknown parent block. Conode could be out of sync.")
 			return
 		}
+
+		sb.ChildSL = append(sb.ChildSL, pg.Genesis.Hash)
+		s.db.Store(sb)
 	}
-	_, err := s.db.StoreBlocks(sbs.SkipBlocks)
-	if err != nil {
-		log.Error(err)
+}
+
+// PropagateForwardLinkHandler will update the latest block with
+// the new forward link and the new block when given
+func (s *Service) propagateForwardLinkHandler(msg network.Message) {
+	pfl, ok := msg.(*PropagateForwardLink)
```

Comment:
We need to think about security here... pfl is essentially untrusted input from the network at this point. Anyone who manages to send the right message into onet can get us to execute this code with their input. What bad stuff can they do? (This was already true before, but I'm wondering if somehow this change opens us up to trying SyncChains we shouldn't be trying?)

Reply:
The roster could actually be taken from the previous blocks. That would fix the three points you pointed out.